src/Security/MemberAuthenticator/ChangePasswordHandler.php

```diff
@@ -348,4 +403,40 @@ protected function checkPassword($member, $password)
         }
         return true;
     }
+
+    /**
+     * Generate a temporary auto login hash for the member
+     */
+    private function createAutoLoginTempHash(): string
+    {
+        $member = null;
+        $autoLoginTempHash = '';
+        do {
+            $autoLoginTempHash = Injector::inst()->get(RandomGenerator::class)->randomToken('sha256');
+            $member = Member::get()->find('AutoLoginTempHash', $autoLoginTempHash);
+        } while ($member !== null);
+        return $autoLoginTempHash;
+    }
+
+    /**
+     * Prepare a friendly message for if the login token is invalid or expired.
+     */
+    private function getInvalidTokenResponse(): array
+    {
+        return [
+            'Content' => DBField::create_field(
+                'HTMLFragment',
+                _t(
+                    Security::class . '.NOTERESETLINKINVALID',
+                    '<p>The password reset link is invalid or expired.</p>'
+                    . '<p>You can request a new one <a href="{link1}">here</a> or change your password after'
+                    . ' you <a href="{link2}">log in</a>.</p>',
+                    [
+                        'link1' => Security::lost_password_url(),
+                        'link2' => Security::login_url(),
+                    ]
+                )
+            ),
+        ];
+    }
```

Comment:
This returns a message that used to only be returned in one place - but is now used to give consistent message any time the token is found to be invalid or expired at any point during the process.